### PR TITLE
OIDC JWT authenticator

### PR DESCRIPTION
**Pull request by lei-tang**
Please provide a description for what this PR is for: https://github.com/istio/istio/issues/30295
- Implement an OIDC JWT authenticator that supports both JWKS-URI and OIDC discovery. The OIDC JWT authenticator can be configured through the JWT_RULE env variable and when configured, it will be used as an istiod authenticator.
- Add tests for the OIDC JWT authenticator.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: releasenotes/notes/30294.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: feature
+area: security
+issue:
+- 30295
+releaseNotes:
+- |
+  **Added** OIDC JWT authenticator that supports both JWKS-URI and OIDC discovery. The OIDC JWT authenticator can be configured through the JWT_RULE env variable and when configured, it will be used as an istiod authenticator.
```
Comment (ericvn) [inline releasenotes/notes/30294.yaml:8]: comma in the last sentence isn't correct.  Maybe `The OIDC JWT authenticator will be used when configured through the JWT_RULE env variable.

Comment (lei-tang) [reply releasenotes/notes/30294.yaml]: The PR has been revised according to the comment.

**Review by stewartbutler — APPROVED** [review]
LGTM. @lei-tang should this also be cherrypicked to release-1.9?

**Comment by lei-tang** [comment]
@stewartbutler Thanks! I have added the cherrypick/release-1.9 label.

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -314,6 +316,24 @@ func NewServer(args *PilotArgs) (*Server, error) {
 
 	caOpts.Authenticators = authenticators
 	if features.XDSAuth {
+		authenticators := []security.Authenticator{
```
Comment (costinm) [inline pilot/pkg/bootstrap/server.go:319]: Shouldn't the CA authenticator also add the OIDC auth ? I don't think there is any use case where we want a different set.

**Review by costinm — COMMENTED** [review]
Looks great - except the initialization, we had this problem in the past where the auth list for XDS and CA was different.

File: pilot/pkg/bootstrap/server.go (hunk 2 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go:319]: The OIDC auth is added as the CA authenticator in pilot/pkg/bootstrap/istio_ca.go:
		jwtRule := v1beta1.JWTRule{Issuer: iss, Audiences: []string{aud}}
		oidcAuth, err := authenticate.NewJwtAuthenticator(&jwtRule, opts.TrustDomain)

Comment (costinm) [reply pilot/pkg/bootstrap/server.go:319]: So why do we need duplicated code ? I meant: we should have a single code that initializes the list, and use it in both places, else they'll get out of sync and is harder to maintain.

Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go:319]: This PR does not change the existing CA auth flow. We can address the duplicate initialization in a separate PR.

Comment (howardjohn) [reply pilot/pkg/bootstrap/server.go:319]: The authenticators should be exactly the same. please do not make them different.

Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go:319]: Regarding to the review comment "The authenticators should be exactly the same.", the JWT_RULE is for XDS authentication, not for CA. Therefore, the authenticators for XDS and CA are not exactly the same.

Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go]: The PR has been revised based on the review comments such that XDS and CA use the same authenticators.

File: pilot/pkg/bootstrap/server.go [hunk 3]
```diff
@@ -311,11 +313,20 @@ func NewServer(args *PilotArgs) (*Server, error) {
 		&authenticate.ClientCertAuthenticator{},
 		kubeauth.NewKubeJWTAuthenticator(s.kubeClient, s.clusterID, s.multicluster.GetRemoteKubeClient, spiffe.GetTrustDomain(), features.JwtPolicy.Get()),
 	}
-
-	caOpts.Authenticators = authenticators
 	if features.XDSAuth {
+		if args.JwtRule != "" {
```
Comment (costinm) [inline pilot/pkg/bootstrap/server.go:317]: Move this up. If JwtRule is not set, we still want OIDC auth in the chain for CA.

Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go]: This has been moved up.

**Review by costinm — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Pull request merged**